File: EPSILON/SAMPLE FILE/E-publishing/Tools/Altova/Altova/XMLSpy2006/spl/java/altova/text/Generator.java
[call GenerateFileHeader("Generator.java")]
package com.altova.text;

import java.io.FileWriter;
import java.io.Writer;

import com.altova.AltovaException;

public class Generator {
	private ITextNode m_Current = NullTextNode.getInstance();

	private void switchToParent() {
		ITextNode parent = m_Current.getParent();
		if (!parent.isNull())
			m_Current = parent;
	}

	public void init() {
		m_Current = NullTextNode.getInstance();
	}

	public void saveAsSimpleXML(String filename) {
		try {
			Writer writer = new FileWriter(filename);
			TextNodeXMLSerializer xmlSerializer = new TextNodeXMLSerializer(
					writer);
			xmlSerializer.serialize(m_Current.getRoot());
		} catch (Exception e) {
			throw new AltovaException(e.getMessage());
		}
	}

	public void resetToRoot() {
		m_Current = m_Current.getRoot();
	}

	public void enterElement(String name, byte nodeClass) {
		TextNode node = new TextNode(m_Current, name, nodeClass);
		if (m_Current.isNull())
			node = new RootTextNode(node);
		m_Current = node;
	}

	public void leaveElement(String name) {
		if (0 == name.length())
			switchToParent();
		else {
			ITextNode namednode = m_Current.findNodeUpwardsByName(name);
			if (!namednode.isNull()) {
				m_Current = namednode;
				this.switchToParent();
			}
		}
	}

	public void insertElement(String name, String value, byte nodeClass) {
		TextNode node = new TextNode(m_Current, name, nodeClass);
		node.setValue(value);
	}

	public boolean doesNameEqual(String name) {
		return (name == m_Current.getName());
	}

	public boolean doesNamedChildExist(String name) {
		return (0 < m_Current.getChildren().filterByName(name).size());
	}

	public RootTextNode getRootNode() {
		if (m_Current.isNull())
			return null;
		return (RootTextNode) m_Current.getRoot();
	}

	public void setRootNode(ITextNode rhs) {
		m_Current = new RootTextNode(rhs);
	}

	public ITextNode getCurrentNode() {
		return m_Current;
	}
}

File: EPSILON/TOOLS/Altova/Altova/XMLSpy2006/spl/cs/AltovaText/Generator.cs
[call GenerateFileHeader("Generator.cs")]
using System.IO;
namespace Altova.TextParser
{
	/// <summary>
	/// Encapsulates methods needed to build a tree of <see cref="TextNode"/>s,
	/// methods needed by some <see cref="Flex.Condition"/>s or <see cref="EDIFACT.ICondition"/>s and methods
	/// for saving the created tree to an XML file.
	/// </summary>
	public class Generator
	{
		#region Implementation Detail:
		ITextNode mCurrent = NullTextNode.Instance;

		void SwitchToParent()
		{
			if (!mCurrent.Parent.IsNull) mCurrent = mCurrent.Parent;
		}
		#endregion
		#region Public Interface:
		/// <summary>
		/// Initializes the generator.
		/// </summary>
		public void Init()
		{
			mCurrent = NullTextNode.Instance;
		}
		/// <summary>
		/// Resets the generator to the root node of the tree.
		/// </summary>
		public void ResetToRoot()
		{
			mCurrent = mCurrent.Root;
		}
		/// <summary>
		/// Adds a new node to the children of the current node and tries to assign it the specified name.
		/// </summary>
		/// <param name="name">the name to assign</param>
		/// <param name="eClass">the name of the node's class</param>
		/// <remarks>
		/// If the new node doesn't already have a name assigned, the current node will be set to
		/// the newly created.
		/// </remarks>
		public void EnterElement(string name, NodeClass eClass)
		{
			TextNode node = new TextNode(mCurrent, name, eClass);
			if (mCurrent.IsNull) node = new RootTextNode(node);
			mCurrent = node;
		}
		/// <summary>
		/// Leaves the current node by climbing upwards in the hierarchy.
		/// </summary>
		/// <param name="name">the name of the node to leave</param>
		/// <remarks>
		/// If name is an empty string, this instance climbs only one node.
		/// Otherwise, it searches the next node upwards the hierarchy matching the name.
		/// If found, it climbs to the <see cref="TextNode.Parent"/> of this node, otherwise
		/// it doesn't climb at all.
		/// </remarks>
		public void LeaveElement(string name)
		{
			// if the name is empty, just leave the current element, otherwise
			// move upwards inside the tree to the parent of the node with the specified name. 
			if (0 == name.Length) this.SwitchToParent();
			else
			{
				ITextNode namednode = mCurrent.SearchUpwardsByName(name);
				if (!namednode.IsNull)
				{
					mCurrent = namednode;
					this.SwitchToParent();
				}
			}
		}
		/// <summary>
		/// Inserts a new node into the hierarchy as the child of the current node and 
		/// assigns it the specified name and value.
		/// </summary>
		/// <param name="name">the name to assign</param>
		/// <param name="val">the value to assign</param>
		/// <param name="eClass">the name of the node's class</param>
		public void InsertElement(string name, string val, NodeClass eClass)
		{
			TextNode node = new TextNode(mCurrent, name, eClass);
			node.Value = val;
		}
		/// <summary>
		/// Checks whether the current node's name property equals the specified name.
		/// </summary>
		/// <param name="name">the name to compare to</param>
		/// <returns>true if the name matches, otherwise false</returns>
		public bool DoesNameEqual(string name)
		{
			return (name == mCurrent.Name);
		}
		/// <summary>
		/// Checks whether there is a node among the current node's children whose name
		/// matches a specified parameter. 
		/// </summary>
		/// <param name="name">the name to match</param>
		/// <returns>true if a child with matching name was found, otherwise false</returns>
		public bool DoesNamedChildExist(string name)
		{
			return (0 < mCurrent.Children.FilterByName(name).Length);
		}
		/// <summary>
		/// Gets the root node of the contained hierarchy. May be null.
		/// </summary>
		public RootTextNode RootNode
		{
			get
			{
				return mCurrent.Root as RootTextNode;
			}
		}
		/// <summary>
		/// Sets the root node of the contained hierarchy.
		/// </summary>
		/// <param name="rhs"></param>
		public void SetRootNode(ITextNode rhs)
		{
			mCurrent = new RootTextNode(rhs);
		}
		/// <summary>
		/// Gets the current node of the contained hierarchy. Maybe null.
		/// </summary>
		public ITextNode CurrentNode
		{
			get
			{
				return mCurrent;
			}
		}
		#endregion
		#region Internal interface:
		internal void Dump(string filename)
		{
			using (StreamWriter stream= new StreamWriter(filename))
			{
				TextNodeVisitorXML dumper= new TextNodeVisitorXML(stream);
				dumper.Serialize(RootNode);
			}
		}
		#endregion
	}
}